packages/core/src/fusion/Map.js

```diff
@@ -25,6 +26,10 @@ export default class Map {
    * @returns {Map|FilterMap} mapped source, possibly fused
    */
   static create (f, source) {
+    if (isCanonicalEmpty(source)) {
+      return source
```

Comment:
**tl;dr** I think returning `empty()` is the right thing here, and will be the right thing in many cases.

Thinking a bit more about this `isCanonicalEmpty` vs singleton business we've talked about on slack, I think letting types guide us can be helpful, even though we don't have types in the source yet.  If we think about the type of `map`:

```hs
map :: (a -> b) -> Stream a -> Stream b
```

The type of `source` is `Stream a`, and the return type of `map` is `Stream b`.  That means that, _in the general case_, `source`, which is of type `Stream a` can't be a valid return type of `map`, which returns `Stream b` (`source` can be a valid return value, for example, when `a` and `b` are the same concrete type, but that's a specific case, not the general one).

I think that argument is slightly loose right now, because the type of `empty` is currently documented as:

```hs
empty :: () -> Stream *
```

I'm not sure it's really clear what we meant by `*`.  For example, did we mean `any`, or did we mean "inferred"?  The [flow defs say `any`](https://github.com/mostjs/core/blob/master/packages/core/src/index.js.flow#L17).

So returning `empty()` will be ok: forall `b`, `any` is a valid `b`.  It probably will turn out to be the right thing in some others cases (`ap` comes to mind), but not all (it'd be possible to return `source` from `filter`, for example, because the types line up).

I realize there are probably spots in existing code that return the upstream source when they shouldn't.  Hopefully we can fix those as we go.

### Aside: strongly typed empty()

I don't think we need to do this now, but, we could look to Haskell's `List`, and it's empty variant for guidance:

```hs
data List a = Nil | Cons a (List a)
```

That means that `Nil` _still has the type_ `List a`.  Empty lists are still strongly typed, and you cannot pass an empty `List Int` to a function that accepts `List String`.

We may want to consider tightening the type of `empty` to:

```hs
empty :: () -> Stream a
```

A type system should infer `a` from what is being demanded by the caller (Flow does this.  I'm not sure about TS, but I suspect it does as well, given enough surrounding type hints).  I know the _implementation_ of `empty` currently uses a singleton, but like we talked about in slack, I feel like that really is an implementation detail.  For example, I can imagine an implementation of `empty` that returns a `new Empty()` on each call, and an implementation of `isCanonicalEmpty` that returns `instanceof Empty`.